### PR TITLE
add Experiment manager tools 

### PR DESCRIPTION
**Pull request by minghangli-uni**
(Expts_manager.py and Expts_manager.yaml) to automate the setup and management of ACCESS-OM3 experiments.

See https://github.com/COSIMA/om3-scripts/issues/33

### CODE REVIEW

File: expts_manager/Expts_manager.py [hunk 1]
```diff
@@ -0,0 +1,1075 @@
...
+
+class Expts_manager(object):
+    """
+    A class to manage ACCESS-OM3 experiment runs, including control and perturbation experiments.
+    Attributes:
+        MOM_prefix (str): Prefix for MOM6 parameters.
+        nml_suffix (str): Suffix for namelist parameters.
+        runseq_prefix (str): Prefix for the coupling timestep in `nuopc.runseq`.
+        combo_suffix (str): Suffix for combo perturbation experiments, i.e., multiple-parameter tests.
+        branch_perturb (str): user-defined branch name for the perturbation.
```
Comment (minghangli-uni) [inline expts_manager/Expts_manager.py:77]:
```suggestion
        branch_perturb (str): branch name for the perturbation.
```

File: expts_manager/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+### Components:
+1. `Expts_manager.py`:
+ - This file contains the `ExptManager` class, which faciliates the setup, configuration, and execution of experiments. Key functionalities include:
+    - Configuration management based on user-defined parameters.
+    - Automation of directory creation, ensuring a smooth workflow for running experiments.
+    - Support for parameter updates, including:
+        - MOM6 for **ACCESS-OM3**,
+        - Fortran namelists for both **ACCESS-OM2** and **ACCESS-OM3**,
+        - Coupling timesteps (`nuopc.runseq`) for **ACCESS-OM3**,
+        - Component settings (`nuopc.runconfig`) **ACCESS-OM3**.
```
Comment (aekiss) [inline expts_manager/README.md:24]:
```suggestion
        - Component settings (`nuopc.runconfig`) for **ACCESS-OM3**.
```

File: expts_manager/Expts_manager.yaml [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+MOM_input:
+    HFREEZE: 12
+    BOUND_SALINITY: False
+
+
+# ============ Namelist Tunning ================================
+# Tune parameters across different model components.
+
+# Generlised structure
```
Comment (aekiss) [inline expts_manager/Expts_manager.yaml:103]:
```suggestion
# Generalised structure
```

File: expts_manager/write_Expts_manager_yaml.py [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+            },
+        },
+    ]
+
+    # Descriptions and configs for Namelist Tunning
+    descrpt_namelists = """
+# ============ Namelist Tunning ================================
+# Tune parameters across different model components.
+
+# Generlised structure
```
Comment (aekiss) [inline expts_manager/write_Expts_manager_yaml.py:257]:
```suggestion
# Generalised structure
```

File: expts_manager/Expts_manager.yaml [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+# Adjust this value to set the number of iterations for the control experiment, which serves as the baseline for perturbations.
+
+run_namelists: True  
+# Determines whether to run using the specified namelists.            
+# Default: False.            
+# Set to 'True' to apply configurations from the namelist section; otherwise, 'False' to skip this step.
+
+check_duplicate_jobs: True  
+# Checks if there are duplicate PBS jobs within the same parent directory (`test_path`) based on their names.            
+# This check is useful when you have existing running jobs and want to add additional tests, which helps avoid conflicts by ensuring new jobs dont duplicate exisiting ones in the same `test_path`.
```
Comment (aekiss) [inline expts_manager/Expts_manager.yaml:262]:
```suggestion
# This check is useful when you have existing running jobs and want to add additional tests, which helps avoid conflicts by ensuring new jobs dont duplicate existing ones in the same `test_path`.            
```

File: expts_manager/write_Expts_manager_yaml.py [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+            "value": True,
+            "comment": "\n# Determines whether to run using the specified namelists.\
+            \n# Default: False.\
+            \n# Set to 'True' to apply configurations from the namelist section; otherwise, 'False' to skip this step.\n",
+        },
+        {
+            "key": "check_duplicate_jobs",
+            "value": True,
+            "comment": "\n# Checks if there are duplicate PBS jobs within the same parent directory (`test_path`) based on their names.\
+            \n# This check is useful when you have existing running jobs and want to add additional tests, which helps avoid conflicts by ensuring new jobs dont duplicate exisiting ones in the same `test_path`.\
```
Comment (aekiss) [inline expts_manager/write_Expts_manager_yaml.py:436]:
```suggestion
            \n# This check is useful when you have existing running jobs and want to add additional tests, which helps avoid conflicts by ensuring new jobs don't duplicate existing ones in the same `test_path`.\
```

File: expts_manager/write_Expts_manager_yaml.py [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+            \n# The check will not be triggered if the jobs are located in different `test_path`. It only applies to jobs within the same `test_path` directory.\
+            \n# Default: True.\
+            \n# If duplicates are found, a message will be printed indicating the presence of duplicate runs and those runs will be skipped.\n",
+        },
+        {
+            "key": "check_skipping",
+            "value": False,
+            "comment": "\n# Checks if certain runs should be skipped based on pre-existing conditions or results. Currently only valids to nml type.\
+            \n# Default: False.\
+            \n# Set to 'True' if you want the system to skip runs under specific criteria; otherwise, keep it 'False'. Currently only valids to nml type.\n",
```
Comment (aekiss) [inline expts_manager/write_Expts_manager_yaml.py:446]:
```suggestion
            "comment": "\n# Checks if certain runs should be skipped based on pre-existing conditions or results. Currently only valid for nml type.\
            \n# Default: False.\
            \n# Set to 'True' if you want the system to skip runs under specific criteria; otherwise, keep it 'False'. Currently only valid for nml type.\n",
```

File: expts_manager/Expts_manager.yaml [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+check_duplicate_jobs: True  
+# Checks if there are duplicate PBS jobs within the same parent directory (`test_path`) based on their names.            
+# This check is useful when you have existing running jobs and want to add additional tests, which helps avoid conflicts by ensuring new jobs dont duplicate exisiting ones in the same `test_path`.            
+# The check will not be triggered if the jobs are located in different `test_path`. It only applies to jobs within the same `test_path` directory.            
+# Default: True.            
+# If duplicates are found, a message will be printed indicating the presence of duplicate runs and those runs will be skipped.
+
+check_skipping: False  
+# Checks if certain runs should be skipped based on pre-existing conditions or results. Currently only valids to nml type.
```
Comment (aekiss) [inline expts_manager/Expts_manager.yaml:268]:
```suggestion
# Checks if certain runs should be skipped based on pre-existing conditions or results. Currently only valid for nml type.            
```

File: expts_manager/Expts_manager.yaml [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+# Checks if there are duplicate PBS jobs within the same parent directory (`test_path`) based on their names.            
+# This check is useful when you have existing running jobs and want to add additional tests, which helps avoid conflicts by ensuring new jobs dont duplicate exisiting ones in the same `test_path`.            
+# The check will not be triggered if the jobs are located in different `test_path`. It only applies to jobs within the same `test_path` directory.            
+# Default: True.            
+# If duplicates are found, a message will be printed indicating the presence of duplicate runs and those runs will be skipped.
+
+check_skipping: False  
+# Checks if certain runs should be skipped based on pre-existing conditions or results. Currently only valids to nml type.            
+# Default: False.            
+# Set to 'True' if you want the system to skip runs under specific criteria; otherwise, keep it 'False'. Currently only valids to nml type.
```
Comment (aekiss) [inline expts_manager/Expts_manager.yaml:270]:
```suggestion
# Set to 'True' if you want the system to skip runs under specific criteria; otherwise, keep it 'False'. Currently only valid for nml type.
```

File: expts_manager/Expts_manager.py [hunk 10]
```diff
@@ -0,0 +1,1427 @@
...
+        if self.diag_ctrl and self.diag_path:
+            if self.model == "access-om3":
+                self._copy_diag_table(base_path)
+            elif self.model == "access-om2":
+                self._copy_diag_table(ocn_path)
+
+        # setup the control experiments
+        self._setup_ctrl_expt()
+
+        # check exisiting pbs jobs
```
Comment (aekiss) [inline expts_manager/Expts_manager.py:295]:
```suggestion
        # check existing pbs jobs
```

File: expts_manager/Expts_manager.py [hunk 11]
```diff
@@ -0,0 +1,1427 @@
...
+
+    def _generate_restart_symlink(self, expt_path):
+        """
+        Generates a symlink to the restart directory if needed.
+
+        Args:
+            expt_path (str): The path to the experiment directory.
+        """
+        if self.startfrom_str != "rest":
+            # if exisiting restarts and not force generating restart, then skip.
```
Comment (aekiss) [inline expts_manager/Expts_manager.py:1026]:
```suggestion
            # if existing restarts and not force generating restart, then skip.
```

**Review by aekiss — APPROVED** [review]
